variable.c

```diff
@@ -1440,15 +1441,24 @@ rb_obj_ivar_set(VALUE obj, ID id, VALUE val)
 
         RUBY_ASSERT(!rb_shape_obj_too_complex(obj));
 
-        if (UNLIKELY(shape->next_iv_index >= num_iv)) {
+        VALUE klass = rb_obj_class(obj);
+        bool too_complex = RCLASS_EXT(klass)->variation_count >= SHAPE_MAX_VARIATIONS;
+
+        /* Grow capacity of ivar list. Don't grow capa it if we're about to
+           switch to OBJ_TOO_COMPLEX shape */
+        if (UNLIKELY(!too_complex && shape->next_iv_index >= num_iv)) {
             RUBY_ASSERT(shape->next_iv_index == num_iv);
 
             shape = rb_grow_iv_list(obj);
             RUBY_ASSERT(shape->type == SHAPE_CAPACITY_CHANGE);
         }
 
         rb_shape_t *next_shape = rb_shape_get_next(shape, obj, id);
```

Comment:
I think the code would be easier to read if we extracted the check for `too_complex` out of `rb_shape_get_next` and instead perform it here instead since this is the only place that calls `rb_shape_get_next` with a `T_OBJECT`.

Reply:
Okay I'll do that, thanks for the review!

Reply:
I see that makes sense.